src/pages/patientView/timeline2/VAFChartWrapper.tsx

```diff
@@ -42,108 +61,424 @@ export interface IVAFChartWrapperProps {
+            .flatten()
+            .map((d: IPoint) => d.y)
+            .max();
+    }
 
-            (window as any).store = store1;
-        }, []);
+    @computed get sampleIdToClinicalValue() {
+        let sampleIdToClinicalValue: { [sampleId: string]: string } = {};
+        if (this.wrapperStore.groupingByIsSelected) {
+            this.props.sampleManager.samples.forEach((sample, i) => {
+                sampleIdToClinicalValue[
+                    sample.id
+                ] = SampleManager!.getClinicalAttributeInSample(
+                    sample,
+                    this.wrapperStore.groupByOption!
+                )!.value;
+            });
+        }
+        return sampleIdToClinicalValue;
+    }
+
+    @computed get sampleIconsTracks() {
+        const tracks: CustomTrackSpecification[] = [];
+        if (this.wrapperStore.groupingByIsSelected) {
+            _.forIn(this.sampleGroups, (sampleIds: string[], key: string) => {
+                const index = parseInt(key);
+                tracks.push({
+                    renderHeader: () => this.groupByTrackLabel(index),
+                    renderTrack: () => this.sampleIcons(sampleIds),
+                    height: () => 20,
+                    labelForExport: this.clinicalValuesForGrouping[index],
+                });
+            });
+        } else {
+            tracks.push({
+                renderHeader: () => '',
+                renderTrack: () => this.sampleIcons(this.store.sampleIds),
+                height: () => 20,
+                labelForExport: 'VAF Samples',
+            });
+        }
+        return tracks;
+    }
 
-        if (!stores || !wrapperStore) return null;
+    @computed get xPosition() {
+        let positionList: { [sampleId: string]: number } = {};
+        let sequentialDistance: number = 0;
+        let sequentialPadding: number = 20;
+        if (this.wrapperStore.showSequentialMode) {
+            sequentialDistance =
+                (this.store.pixelWidth - sequentialPadding * 2) /
+                (this.store.sampleEvents.length - 1);
+        }
 
-        const groupByTracks = wrapperStore.groupByTracks;
+        this.store.sampleEvents.forEach((sample, i) => {
+            sample.event.attributes.forEach((attribute: any, i: number) => {
+                if (attribute.key === 'SAMPLE_ID') {
+                    positionList[attribute.value] = this.wrapperStore
+                        .showSequentialMode
+                        ? this.sampleIdOrder[attribute.value] *
+                              sequentialDistance +
+                          sequentialPadding
+                        : this.store.getPosition(sample)!.pixelLeft;
+                }
+            });
+        });
+        return positionList;
+    }
+
+    @computed get sampleIdOrder() {
+        return stringListToIndexSet(
+            this.props.sampleManager.getSampleIdsInOrder()
+        );
+    }
+
+    groupColor(sampleId: string) {
```

Comment:
should be computed

Reply:
Why?

Reply:
BTW can function, not getter, be computed?

Reply:
I guess because you need to trigger recalculation when some of the variables `groupColor` rely on change.
I am wondering when one would not choose to make something `@computed`? When it's a pure function that only depends on it's input?